src/testrender/shading.cpp

```diff
@@ -192,13 +393,13 @@ struct OrenNayar final : public BSDF, OrenNayarParams {
             float LV    = -sg.I.dot(wi);
             float s     = LV - NL * NV;
             float stinv = s > 0 ? s / std::max(NL, NV) : 0.0f;
-            return A + B * stinv;
+            return Color3(A + B * stinv);
         }
-        return pdf = 0;
+        return Color3(pdf = 0);
     }
-    virtual float sample(const OSL::ShaderGlobals& sg, float rx, float ry,
+    Color3 sample(const OSL::ShaderGlobals& sg, float rx, float ry,
                          float /*rz*/, OSL::Dual2<OSL::Vec3>& wi,
-                         float& pdf) const
+                         float& pdf) const override
```

Comment:
This patch looks good to me, thanks! But I'm going to abuse it to open a discussion on the BSDF API:

  1) Should we return a struct like { weight, pdf } to avoid output args?
  2) Can we drop the full sg as argument and instead pass a Vec3 wo (meaning I) so you can evaluate them in an SG-less context?
  3) Does anybody need accurate derivs from wi? If not, can we stop pretending we can compute them?
  4) Should we add roughness to the output? I believe most renders track it. You can guess it from pdf following a convention (like reversing an isotropic GGX), but wanted to get opinions. A normalized [0-1] roughness can be used as a measure of specularity for AOVs. If we sign off on BSDFs being single lobe the roughness can be just a BSDF property and not an output from sample()/eval().

I volunteer to making any of these changes after you merge this, of course.

Reply:
1. Yes - good idea. I can do that refactor in another PR. Would you have a { weight, pdf, dir } struct for sample() then?
2. Yes - good idea.
3. Yes - definitely. Only the reflection() and refraction() cases do anything sensible. Do you want to submit a patch that switches to ray cones? That would be a nice simplification for the bsdf API.
4. I wouldn't bother with this for now. I prefer keeping testrender minimal for now.

Reply:
Yes { color, wi, pdf } in any order you like works for me. I will do the cone tracing.